tt_metal/graph_tracking.hpp

```diff
@@ -0,0 +1,137 @@
+            REAL,
+            FAKE
+        };
+
+        IGraphProcessor() = default;
+
+        virtual void track_allocate(tt::tt_metal::Buffer* buffer, bool bottom_up) {};
+
+        virtual void track_deallocate(tt::tt_metal::Buffer* buffer) {};
+
+        virtual void track_allocate_cb(const CoreRangeSet &core_range_set, uint64_t addr, uint64_t size) {};
+
+        virtual void track_deallocate_cb() {};
+
+        virtual void track_program(tt::tt_metal::Program* program) {};
+
+        virtual void track_begin_op(std::string_view function_name, std::span<std::any> input_parameters) {};
+
+        virtual void track_end_op() {};
+        virtual void track_end_op(const std::any& output_tensors) {};
+
+        virtual void begin_capture(RunMode mode) {};
+
+        virtual std::string end_capture() {return "";};
+
+        virtual ~IGraphProcessor() = default;
+
+    };
+
+    class IGraphHooks {
+    public:
+        IGraphHooks() = default;
+        virtual bool hook_allocate(tt::tt_metal::Buffer* buffer, bool bottom_up) = 0;
+
+        virtual bool hook_deallocate(tt::tt_metal::Buffer* buffer) = 0;
+
+        virtual bool hook_program(Program* program) = 0;
+
+        virtual ~IGraphHooks() = default;
+    };
+
+    class GraphTracker {
+    public:
+        static GraphTracker& instance() {
+            static GraphTracker tracker;
+            return tracker;
+        }
+
+        void push_processor(const std::shared_ptr<IGraphProcessor>& processor);
+        void pop_processor();
+
+        bool add_hook(const std::shared_ptr<IGraphHooks>& hook);
+
+        void track_allocate(Buffer* buffer, bool bottom_up);
+
+        void track_deallocate(Buffer* buffer);
+
+        void track_allocate_cb(const CoreRangeSet &core_range_set, uint64_t addr, uint64_t size);
+
+        void track_deallocate_cb();
+
+        void track_program(Program* program);
+
+        template<class... Args>
+        void track_begin_op(std::string_view function_name, Args&&... args) {
+            if (processors.empty()) {
+                return;
+            }
+            // we don't know anything in metal about ttnn classes so std::any is the only option here
+            std::array<std::any, sizeof...(Args)>  params{std::any(std::ref(args))...};
+            for (auto& it : processors) {
+                it->track_begin_op(function_name, params);
+            }
+        }
+
+        // Track op that doesn't return anything
+        void track_end_op() {
+            if (processors.empty()) {
+                return;
+            }
+            for (auto& it : processors) {
+                it->track_end_op();
+            }
+        }
+
+        template<class ReturnType>
+        void track_end_op(ReturnType&& output_tensors) {
+            if (processors.empty()) {
+                return;
+            }
+            for (auto& it : processors) {
+                it->track_end_op(std::ref(output_tensors));
+            }
+        }
```

Comment:
Why do we need to pass down ttnn classes into the metal graph fns?
I also thought we wanted to avoid the connotation of ttnn/eager concepts like ops and tensors within tt_metal? @ayerofieiev-tt

Reply:
Great catch, Austin! Thank you for spotting!
@dmakoviichuk-tt , we should discuss. 
Metal is not aware about a concept of an Operation. Thats a TT-NN concept.

Reply:
Even just name track_begin_op/track_end_op is a concept leak

Reply:
I don't see a problem here with passing down classes. It is pretty common approach when you define some interface in a low level and it implemented later in highlevel part.
@ayerofieiev-tt  we can just rename it to the track_begin_function/track_end_function.

Reply:
Agreed

Reply:
Also I see that output_tensors is a bad name too.
Should be just output.

Reply:
@tt-aho good catch with names. Actually it is just function tracking. We will rename it to be less confused.